165055215 Implement Users Login Feature

Simplemart17 opened the pull request:
#### Description
- This PR add function that enables users to login using their email and password
#### Type of change
- New feature (non-breaking change which add functionality)
#### How Has This Been Tested?
- Integration
#### Checklist:
- My code follows the style guideline of this project
#### PT
[#165055215](https://www.pivotaltracker.com/story/show/165055215)
#### Screencast
![login](https://user-images.githubusercontent.com/41796727/55667118-2604fd80-5850-11e9-8942-852b12ca5b77.gif)

Simplemart17 commented on src/controllers/users.js:
```diff
@@ -0,0 +1,33 @@
+          email,
+          password,
+        },
+      }).then(users => {
```
fixed.

Temmyogunbo commented on src/controllers/users.js:
```diff
@@ -0,0 +1,32 @@
+          data: userDetails,
+        });
+      }
+    } catch (error) {
```
we can use `next` middleware here.

Temmyogunbo commented on src/config/passportConfig.js:
```diff
@@ -0,0 +1,38 @@
+        if (!user) {
+          return done(null, false, { message: 'Incorrect credentials' });
+        }
+        if (password !== user.password) {
```
Just so you know, we will not be storing a plain password in the db. Seed the database with a hashed password and use the compare method we attached to the user model prototype. This will always fail when local signup is ready

Temmyogunbo commented on src/config/passportConfig.js:
```diff
@@ -0,0 +1,38 @@
+          return done(null, false, { message: 'Incorrect credentials' });
+        }
+        if (password !== user.password) {
+          return done(null, false, { message: 'Incorrect credentials' });
```
sync with @chikeozulumba and @Raymond-Osy for consistent error format.

Temmyogunbo commented on tests/userLogin.test.js:
```diff
@@ -0,0 +1,29 @@
+        password: '%RYYT&^UTB*UYT*IUYIU',
+      })
+      .end((err, res) => {
+        expect(res).to.have.status(200);
```
Sync with @chikeozulumba for better testing.

Simplemart17 replied:
I have written more test. Thank you.

Simplemart17 committed:
165055215-feature(loginAPI): add function for user to login

- add function to authenticate user using passport

[Delivers #165055215]

Temmyogunbo commented on tests/integration/userLogin.test.js:
```diff
@@ -0,0 +1,53 @@
+        password: '%RYYT&^UTB*UYT*IUYIU',
+      })
+      .end((err, res) => {
+        expect(res).to.have.status(200);
```
You need to test further. Test that email and password is the same.

Temmyogunbo commented on src/models/user.js:
```diff
@@ -102,5 +105,9 @@ export default (sequelize, DataTypes) => {
     });
   };
+
+  User.hashPassword = password => bcrypt.hashSync(password, bcrypt.genSaltSync(10));
```
You can remove this. @Raymond-Osy has proper implementation.

Temmyogunbo replied:
You can relax on this. If I merge @Raymond-Osy work first then you gonna remove it.

Simplemart17 committed:
165055215-feature(loginAPI): add more test for user login

- add test to decode login token

[Delivers #165055215]

Simplemart17 committed:
165055215-feature(loginAPI): refactor user login

- refactor user login feature

[Delivers #165055215]

Simplemart17 committed:
165055215-feature(loginAPI): users login feature

- modify validation middleware

[Delivers #165055215]

Temmyogunbo commented on src/middlewares/localStrategy.js:
```diff
@@ -0,0 +1,39 @@
+        if (!user) {
+          return done(null, false);
+        }
+        if (!Auth.comparePassword(password, user.password)) {
```
Please put this method inside prototype object of user model.

Simplemart17 committed:
165055215-feature(loingAPI): refactor user login

- add error response format function
- add more test cases

[Delivers #165055215]
